nexus/db-queries/src/db/datastore/target_release.rs

```diff
@@ -0,0 +1,265 @@
+    /// existing row.
+    pub async fn target_release_insert(
+        &self,
+        opctx: &OpContext,
+        target_release: TargetRelease,
+    ) -> CreateResult<TargetRelease> {
+        opctx
+            .authorize(authz::Action::Modify, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+        self.transaction_retry_wrapper("set_target_release")
+            .transaction(&conn, |conn| {
+                let target_release = target_release.clone();
+                async move {
+                    insert_into(dsl::target_release)
+                        .values(target_release)
+                        .returning(TargetRelease::as_returning())
+                        .get_result_async(&conn)
+                        .await
+                }
+            })
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Convert a model-level target release to an external view.
+    /// This method lives here because we have to look up the version
+    /// corresponding to the TUF repo.
+    pub async fn target_release_view(
+        &self,
+        opctx: &OpContext,
+        target_release: &TargetRelease,
+    ) -> LookupResult<views::TargetRelease> {
+        opctx
+            .authorize(authz::Action::Read, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+        Ok(views::TargetRelease {
+            generation: (&target_release.generation.0).into(),
+            time_requested: target_release.time_requested,
+            release_source: match target_release.release_source {
+                TargetReleaseSource::Unspecified => {
+                    views::TargetReleaseSource::Unspecified
+                }
+                TargetReleaseSource::SystemVersion => {
+                    use crate::db::schema::tuf_repo;
+                    if let Some(tuf_repo_id) = target_release.tuf_repo_id {
+                        views::TargetReleaseSource::SystemVersion(
+                            tuf_repo::table
+                                .select(tuf_repo::system_version)
+                                .filter(tuf_repo::id.eq(tuf_repo_id))
+                                .first_async::<SemverVersion>(&*conn)
+                                .await
+                                .map_err(|e| {
+                                    public_error_from_diesel(
+                                        e,
+                                        ErrorHandler::Server,
+                                    )
+                                })?
+                                .into(),
+                        )
+                    } else {
+                        return Err(Error::invalid_request(
+                            "missing TUF repo ID for specified system version",
+                        ));
```

Comment:
In practice, is this likely a case where the system version is currently set to a TUF repo that we don't actually have?  If that's the case, I feel like this should be a 500 (server error), not a 400 -- we shouldn't allow this to happen.  (It's not this PR's job to prevent it, to be clear.)

But if this could be called with system version's that aren't the current one, where we might reasonably have deleted the old TUF repo, then maybe it needs to return an error type that would allow the caller to decide whether it's a 500 or not.

Reply:
Thanks, fixed in 8e3301482.